File: internal/crypto/dy/ab/sign.go
package ab

import (
	"crypto/rc4"
	"encoding/hex"
	"fmt"
	"math/rand"
	"strings"
	"time"

	"github.com/tjfoc/gmsm/sm3"
)

var sObj = map[string]string{
	"s0": "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=",
	"s1": "Dkdpgh4ZKsQB80/Mfvw36XI1R25+WUAlEi7NLboqYTOPuzmFjJnryx9HVGcaStCe=",
	"s2": "Dkdpgh4ZKsQB80/Mfvw36XI1R25-WUAlEi7NLboqYTOPuzmFjJnryx9HVGcaStCe=",
	"s3": "ckdp1h4ZKsUB80/Mfvw36XIgR25+WQAlEi7NLboqYTOPuzmFjJnryx9HVGDaStCe",
	"s4": "Dkdpgh2ZmsQB80/MfvV36XI1R45-WUAlEixNLwoqYTOPuzKFjJnry79HbGcaStCe",
}

// Signer holds configuration for deterministic signing.
type Signer struct {
	fixedTimestamp int64
	randomValues   []float64
	randIdx        int
}

// NewSigner creates a new Signer for deterministic signing.
func NewSigner(fixedTimestamp int64, randomValues []float64) *Signer {
	return &Signer{
		fixedTimestamp: fixedTimestamp,
		randomValues:   randomValues,
	}
}

func (s *Signer) nextRandom() float64 {
	if s.randomValues != nil && len(s.randomValues) > 0 {
		val := s.randomValues[s.randIdx%len(s.randomValues)]
		s.randIdx++
		return val
	}
	return rand.Float64() * 10000
}

func rc4Encrypt(plaintext, key []byte) ([]byte, error) {
	cipher, err := rc4.NewCipher(key)
	if err != nil {
		return nil, err
	}
	dst := make([]byte, len(plaintext))
	cipher.XORKeyStream(dst, plaintext)
	return dst, nil
}

// ResultEncrypt 使用自定义字母表对字符串进行 Base64 编码
// 这个实现逻辑上与 Python 版本完全一致
func ResultEncrypt(longStr string, numKey string) (string, error) {
	// 1. 根据 key 获取字母表
	alphabet, ok := sObj[numKey]
	if !ok {
		return "", fmt.Errorf("invalid numKey: %s", numKey)
	}

	// 2. 将输入字符串转换为字节切片
	// Go 的 string 内部是 UTF-8，直接转换为 []byte 可以逐字节处理，
	// 这与 Python 的 .encode('latin-1') 在处理 ASCII 和单字节字符时行为一致。
	longStrBytes := []byte(longStr)
	dataLen := len(longStrBytes)

	// 3. 使用 strings.Builder 高效地构建结果字符串
	// 预分配容量可以进一步提升性能
	var result strings.Builder
	result.Grow((dataLen/3 + 1) * 4)

	// 4. 以 3 字节为步长遍历数据，这是 Base64 的核心逻辑
	for i := 0; i < dataLen; i += 3 {
		// 定义3个字节变量，并处理数据末尾不足3字节的情况
		// 缺失的字节默认为0，这在位运算中是正确的处理方式
		b1, b2, b3 := longStrBytes[i], byte(0), byte(0)

		if i+1 < dataLen {
			b2 = longStrBytes[i+1]
		}
		if i+2 < dataLen {
			b3 = longStrBytes[i+2]
		}

		// 将3个8位字节（24位）合并成一个 uint32 整数
		// 使用 uint32 确保位移操作不会溢出
		longInt := uint32(b1)<<16 | uint32(b2)<<8 | uint32(b3)

		// 将24位整数拆分为4个6位的索引值
		idx1 := (longInt >> 18) & 0x3F
		idx2 := (longInt >> 12) & 0x3F
		idx3 := (longInt >> 6) & 0x3F
		idx4 := longInt & 0x3F

		// 根据索引从字母表中查找字符并写入结果
		result.WriteByte(alphabet[idx1])
		result.WriteByte(alphabet[idx2])

		// 5. 正确处理补位逻辑
		// 如果原始数据块至少有2个字节，才写入第3个编码字符
		if i+1 < dataLen {
			result.WriteByte(alphabet[idx3])
		}
		// 如果原始数据块有3个字节，才写入第4个编码字符
		if i+2 < dataLen {
			result.WriteByte(alphabet[idx4])
		}
	}

	return result.String(), nil
}

func getLongInt(round int, longStr string) int {
	round *= 3
	if round+2 >= len(longStr) {
		return 0
	}
	return (int(longStr[round]) << 16) | (int(longStr[round+1]) << 8) | int(longStr[round+2])
}

func generRandom(random float64, option []int) []byte {
	return []byte{
		byte((int(random) & 255 & 170) | (option[0] & 85)),
		byte((int(random) & 255 & 85) | (option[0] & 170)),
		byte((int(random) >> 8 & 255 & 170) | (option[1] & 85)),
		byte((int(random) >> 8 & 255 & 85) | (option[1] & 170)),
	}
}

func (s *Signer) generateRc4BbStr(urlSearchParams, userAgent, windowEnvStr, suffix string, arguments []int) (string, error) {
	h := sm3.New()
	h.Write([]byte(urlSearchParams + suffix))
	sum1 := h.Sum(nil)
	h.Reset()
	h.Write(sum1)
	urlSearchParamsList := h.Sum(nil)

	h.Reset()
	h.Write([]byte(suffix))
	sum2 := h.Sum(nil)
	h.Reset()
	h.Write(sum2)
	cus := h.Sum(nil)

	key := []byte{byte(0), 1, byte(arguments[2])}
	// key := []byte{byte(0.00390625), 1, byte(arguments[2])}
	rc4Encrypted, err := rc4Encrypt([]byte(userAgent), key)
	if err != nil {
		return "", err
	}
	uaEncrypted, err := ResultEncrypt(string(rc4Encrypted), "s3")
	if err != nil {
		return "", err
	}
	h.Reset()
	h.Write([]byte(uaEncrypted))
	ua := h.Sum(nil)

	startTime := s.fixedTimestamp
	if startTime == 0 {
		startTime = time.Now().UnixMilli()
	}
	endTime := startTime

	b := make(map[int]int64)
	b[8] = 3
	b[10] = endTime
	b[15] = 6383 // aid
	b[16] = startTime
	b[18] = 44

	b[20] = (b[16] >> 24) & 255
	b[21] = (b[16] >> 16) & 255
	b[22] = (b[16] >> 8) & 255
	b[23] = b[16] & 255
	b[24] = int64(uint64(b[16]) / 256 / 256 / 256 / 256)
	b[25] = int64(uint64(b[16]) / 256 / 256 / 256 / 256 / 256)

	b[26] = int64((arguments[0] >> 24) & 255)
	b[27] = int64((arguments[0] >> 16) & 255)
	b[28] = int64((arguments[0] >> 8) & 255)
	b[29] = int64(arguments[0] & 255)

	b[30] = int64((arguments[1] / 256) & 255)
	b[31] = int64((arguments[1] % 256) & 255)
	b[32] = int64((arguments[1] >> 24) & 255)
	b[33] = int64((arguments[1] >> 16) & 255)

	b[34] = int64((arguments[2] >> 24) & 255)
	b[35] = int64((arguments[2] >> 16) & 255)
	b[36] = int64((arguments[2] >> 8) & 255)
	b[37] = int64(arguments[2] & 255)

	b[38] = int64(urlSearchParamsList[21])
	b[39] = int64(urlSearchParamsList[22])

	b[40] = int64(cus[21])
	b[41] = int64(cus[22])

	b[42] = int64(ua[23])
	b[43] = int64(ua[24])

	b[44] = (b[10] >> 24) & 255
	b[45] = (b[10] >> 16) & 255
	b[46] = (b[10] >> 8) & 255
	b[47] = b[10] & 255
	b[48] = b[8]
	b[49] = int64(uint64(b[10]) / 256 / 256 / 256 / 256)
	b[50] = int64(uint64(b[10]) / 256 / 256 / 256 / 256 / 256)

	pageID := 6241
	b[51] = int64(pageID)
	b[52] = int64((pageID >> 24) & 255)
	b[53] = int64((pageID >> 16) & 255)
	b[54] = int64((pageID >> 8) & 255)
	b[55] = int64(pageID & 255)

	aid := 6383
	b[56] = int64(aid)
	b[57] = int64(aid & 255)
	b[58] = int64((aid >> 8) & 255)
	b[59] = int64((aid >> 16) & 255)
	b[60] = int64((aid >> 24) & 255)

	windowEnvList := []byte(windowEnvStr)
	b[64] = int64(len(windowEnvList))
	b[65] = b[64] & 255
	b[66] = (b[64] >> 8) & 255

	b[69] = 0
	b[70] = b[69] & 255
	b[71] = (b[69] >> 8) & 255

	b[72] = b[18] ^ b[20] ^ b[26] ^ b[30] ^ b[38] ^ b[40] ^ b[42] ^ b[21] ^ b[27] ^ b[31] ^ b[35] ^ b[39] ^ b[41] ^ b[43] ^ b[22] ^
		b[28] ^ b[32] ^ b[36] ^ b[23] ^ b[29] ^ b[33] ^ b[37] ^ b[44] ^ b[45] ^ b[46] ^ b[47] ^ b[48] ^ b[49] ^ b[50] ^ b[24] ^
		b[25] ^ b[52] ^ b[53] ^ b[54] ^ b[55] ^ b[57] ^ b[58] ^ b[59] ^ b[60] ^ b[65] ^ b[66] ^ b[70] ^ b[71]

	var bb []byte
	for _, val := range []int{18, 20, 52, 26, 30, 34, 58, 38, 40, 53, 42, 21, 27, 54, 55, 31, 35, 57, 39, 41, 43, 22, 28, 32, 60, 36, 23, 29, 33, 37, 44, 45, 59, 46, 47, 48, 49, 50, 24, 25, 65, 66, 70, 71} {
		bb = append(bb, byte(b[val]))
	}

	// var aa []int64
	// for _, val := range []int{18, 20, 52, 26, 30, 34, 58, 38, 40, 53, 42, 21, 27, 54, 55, 31, 35, 57, 39, 41, 43, 22, 28, 32, 60, 36, 23, 29, 33, 37, 44, 45, 59, 46, 47, 48, 49, 50, 24, 25, 65, 66, 70, 71} {
	// 	aa = append(aa, (b[val]))
	// }
	// fmt.Println(aa)
	// // 创建一个rune切片
	// runes := make([]rune, len(aa))
	// // 将int类型的码点值赋给rune切片
	// for i, code := range aa {
	// 	runes[i] = rune(code)
	// }
	// // 打印16进制字符串
	// hexStr := ""
	// for i := 0; i < len(runes); i++ {
	// 	hexStr += fmt.Sprintf("%02x", runes[i])
	// }
	// fmt.Println(hexStr)
	// fmt.Println(len(runes))

	bb = append(bb, windowEnvList...)
	bb = append(bb, byte(b[72]))

	encryptedBb, err := rc4Encrypt(bb, []byte{121})
	if err != nil {
		return "", err
	}

	return string(encryptedBb), nil
}

func (s *Signer) generateRandomStr() string {
	var randomStrList []byte
	randomStrList = append(randomStrList, generRandom(s.nextRandom(), []int{3, 45})...)
	randomStrList = append(randomStrList, generRandom(s.nextRandom(), []int{1, 0})...)
	randomStrList = append(randomStrList, generRandom(s.nextRandom(), []int{1, 5})...)
	return string(randomStrList)
}

func (s *Signer) Sign(urlSearchParams, userAgent string, arguments []int) (string, error) {
	rc4BbStr, err := s.generateRc4BbStr(
		urlSearchParams,
		userAgent,
		"1536|747|1536|834|0|30|0|0|1536|834|1536|864|1525|747|24|24|Win32",
		"cus",
		arguments,
	)
	if err != nil {
		return "", err
	}
	resultStr := s.generateRandomStr() + rc4BbStr
	// 打印16进制字符串
	hexStr := ""
	for i := 0; i < len(resultStr); i++ {
		hexStr += fmt.Sprintf("%02x", resultStr[i])
	}
	fmt.Println(hexStr)
	encryptedStr, err := ResultEncrypt(resultStr, "s4")
	if err != nil {
		return "", err
	}
	return encryptedStr + "=", nil
}

func (s *Signer) SignDetail(params, userAgent string) (string, error) {
	return s.Sign(params, userAgent, []int{0, 1, 14})
}

func (s *Signer) SignReply(params, userAgent string) (string, error) {
	return s.Sign(params, userAgent, []int{0, 1, 8})
}

// Top-level functions for backward compatibility

func Sign(urlSearchParams, userAgent string, arguments []int) (string, error) {
	s := NewSigner(0, nil)
	return s.Sign(urlSearchParams, userAgent, arguments)
}

func SignDetail(params, userAgent string) (string, error) {
	s := NewSigner(0, nil)
	return s.SignDetail(params, userAgent)
}

func SignReply(params, userAgent string) (string, error) {
	s := NewSigner(0, nil)
	return s.SignReply(params, userAgent)
}

// Sm3Sum remains a standalone function.
func Sm3Sum(data []byte) string {
	h := sm3.New()
	h.Write(data)
	sum := h.Sum(nil)
	return hex.EncodeToString(sum)
}

File: internal/crypto/dy/ab/sign.py
'''
Python implementation of the sign.js logic.
'''
import time
import math


class SignGenerator:
    '''
    Translates the signing logic from sign.js to Python.
    It can be configured with fixed values for timestamp and random numbers
    to allow for deterministic testing against the JavaScript implementation.
    '''

    def __init__(self, fixed_timestamp=None, random_values=None):
        self._fixed_timestamp = fixed_timestamp
        self._random_values = random_values if random_values is not None else [0.123, 0.456, 0.789]

    def _get_timestamp(self):
        if self._fixed_timestamp is not None:
            return self._fixed_timestamp
        return int(time.time() * 1000)

    def rc4_encrypt(self, plaintext, key):
        s = list(range(256))
        j = 0
        key_bytes = key.encode('latin-1') if isinstance(key, str) else bytes(key)
        plaintext_bytes = plaintext.encode('latin-1') if isinstance(plaintext, str) else bytes(plaintext)

        for i in range(256):
            j = (j + s[i] + key_bytes[i % len(key_bytes)]) % 256
            s[i], s[j] = s[j], s[i]

        i = 0
        j = 0
        cipher = []
        for k in range(len(plaintext_bytes)):
            i = (i + 1) % 256
            j = (j + s[i]) % 256
            s[i], s[j] = s[j], s[i]
            t = (s[i] + s[j]) % 256
            cipher.append(s[t] ^ plaintext_bytes[k])
        return "".join(map(chr, cipher))

    def _le(self, e, r):
        r %= 32
        return ((e << r) | (e >> (32 - r))) & 0xFFFFFFFF

    def _de(self, e):
        if 0 <= e < 16: return 2043430169
        if 16 <= e < 64: return 2055708042
        raise ValueError("invalid j for constant Tj")

    def _pe(self, e, r, t, n):
        if 0 <= e < 16: return (r ^ t ^ n) & 0xFFFFFFFF
        if 16 <= e < 64: return ((r & t) | (r & n) | (t & n)) & 0xFFFFFFFF
        raise ValueError('invalid j for bool function FF')

    def _he(self, e, r, t, n):
        if 0 <= e < 16: return (r ^ t ^ n) & 0xFFFFFFFF
        if 16 <= e < 64: return ((r & t) | (~r & n)) & 0xFFFFFFFF
        raise ValueError('invalid j for bool function GG')

    class SM3:
        def __init__(self, parent):
            self._parent = parent
            self.reg = [0] * 8
            self.chunk = bytearray()
            self.size = 0
            self.reset()

        def reset(self):
            self.reg = [1937774191, 1226093241, 388252375, 3666478592, 2842636476, 372324522, 3817729613, 2969243214]
            self.chunk = bytearray()
            self.size = 0

        def write(self, data):
            if isinstance(data, str):
                a = data.encode('utf-8')
            else:
                a = bytes(data)
            self.size += len(a)
            self.chunk.extend(a)

        def sum(self, data=None, t='bytes'):
            if data is not None:
                self.reset()
                self.write(data)
            
            chunk_copy = self.chunk[:]
            self._fill(chunk_copy)
            
            for i in range(0, len(chunk_copy), 64):
                self._compress(chunk_copy[i:i+64])

            if t == 'hex':
                return "".join([f'{val:08x}' for val in self.reg])
            else:
                i = []
                for val in self.reg:
                    i.extend(val.to_bytes(4, 'big'))
                return i

        def _compress(self, t):
            r = [0] * 68
            for i in range(16):
                r[i] = int.from_bytes(bytes(t[i*4:(i+1)*4]), 'big')

            for n in range(16, 68):
                a = r[n - 16] ^ r[n - 9] ^ self._parent._le(r[n - 3], 15)
                a = a ^ self._parent._le(a, 15) ^ self._parent._le(a, 23)
                r[n] = (a ^ self._parent._le(r[n - 13], 7) ^ r[n - 6]) & 0xFFFFFFFF

            w_prime = [r[j] ^ r[j + 4] for j in range(64)]

            i_reg = self.reg[:]
            for c in range(64):
                o = self._parent._le(i_reg[0], 12) + i_reg[4] + self._parent._le(self._parent._de(c), c)
                o = self._parent._le(o & 0xFFFFFFFF, 7)
                s = (o ^ self._parent._le(i_reg[0], 12)) & 0xFFFFFFFF
                
                u = self._parent._pe(c, i_reg[0], i_reg[1], i_reg[2])
                u = (u + i_reg[3] + s + w_prime[c]) & 0xFFFFFFFF
                
                b = self._parent._he(c, i_reg[4], i_reg[5], i_reg[6])
                b = (b + i_reg[7] + o + r[c]) & 0xFFFFFFFF
                
                i_reg[3] = i_reg[2]
                i_reg[2] = self._parent._le(i_reg[1], 9)
                i_reg[1] = i_reg[0]
                i_reg[0] = u
                i_reg[7] = i_reg[6]
                i_reg[6] = self._parent._le(i_reg[5], 19)
                i_reg[5] = i_reg[4]
                i_reg[4] = (b ^ self._parent._le(b, 9) ^ self._parent._le(b, 17)) & 0xFFFFFFFF

            for l in range(8):
                self.reg[l] = (self.reg[l] ^ i_reg[l]) & 0xFFFFFFFF

        def _fill(self, chunk):
            a = self.size * 8
            chunk.append(128)
            
            while (len(chunk) % 64) != 56:
                chunk.append(0)
            
            chunk.extend(a.to_bytes(8, 'big'))

    def get_sm3_instance(self):
        return self.SM3(self)

    def result_encrypt(self, long_str, num=None):
        s_obj = {
            "s0": "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=",
            "s1": "Dkdpgh4ZKsQB80/Mfvw36XI1R25+WUAlEi7NLboqYTOPuzmFjJnryx9HVGcaStCe=",
            "s2": "Dkdpgh4ZKsQB80/Mfvw36XI1R25-WUAlEi7NLboqYTOPuzmFjJnryx9HVGcaStCe=",
            "s3": "ckdp1h4ZKsUB80/Mfvw36XIgR25+WQAlEi7NLboqYTOPuzmFjJnryx9HVGDaStCe",
            "s4": "Dkdpgh2ZmsQB80/MfvV36XI1R45-WUAlEixNLwoqYTOPuzKFjJnry79HbGcaStCe"
        }
        alphabet = s_obj[num]
        result = []
        long_str_bytes = long_str.encode('latin-1')
        
        for i in range(0, len(long_str_bytes), 3):
            chunk = long_str_bytes[i:i+3]
            b1 = chunk[0]
            b2 = chunk[1] if len(chunk) > 1 else 0
            b3 = chunk[2] if len(chunk) > 2 else 0
            
            long_int = (b1 << 16) | (b2 << 8) | b3

            result.append(alphabet[(long_int >> 18) & 0x3F])
            result.append(alphabet[(long_int >> 12) & 0x3F])
            
            if len(chunk) > 1:
                result.append(alphabet[(long_int >> 6) & 0x3F])
            if len(chunk) > 2:
                result.append(alphabet[long_int & 0x3F])

        return "".join(result)

    def _gener_random(self, random_val, option):
        random_val = int(random_val)
        return [
            (random_val & 255 & 170) | (option[0] & 85),
            (random_val & 255 & 85) | (option[0] & 170),
            ((random_val >> 8) & 255 & 170) | (option[1] & 85),
            ((random_val >> 8) & 255 & 85) | (option[1] & 170),
        ]

    def generate_random_str(self):
        random_str_list = []
        random_str_list.extend(self._gener_random(self._random_values[0] * 10000, [3, 45]))
        random_str_list.extend(self._gener_random(self._random_values[1] * 10000, [1, 0]))
        random_str_list.extend(self._gener_random(self._random_values[2] * 10000, [1, 5]))
        return "".join(map(chr, random_str_list))

    def generate_rc4_bb_str(self, url_search_params, user_agent, window_env_str, suffix="cus", Arguments=[0, 1, 14]):
        start_time = self._get_timestamp()

        # url_search_params_list
        sm3_1 = self.get_sm3_instance()
        sm3_1.write(url_search_params + suffix)
        hash1 = sm3_1.sum()
        url_search_params_list = self.get_sm3_instance().sum(hash1)

        # cus
        sm3_2 = self.get_sm3_instance()
        sm3_2.write(suffix)
        hash2 = sm3_2.sum()
        cus = self.get_sm3_instance().sum(hash2)

        # ua
        rc4_key_bytes = [int(0.00390625), 1, Arguments[2]]
        ua_rc4 = self.rc4_encrypt(user_agent, bytes(rc4_key_bytes))
        ua_encrypted = self.result_encrypt(ua_rc4, "s3")
        ua = self.get_sm3_instance().sum(ua_encrypted)

        end_time = self._get_timestamp()

        b = {}
        b[8] = 3
        b[10] = end_time
        b[15] = {
            "aid": 6383,
            "pageId": 6241,
            "boe": False,
            "ddrt": 7,
            "paths": {
                "include": [
                    {},
                    {},
                    {},
                    {},
                    {},
                    {},
                    {}
                ],
                "exclude": []
            },
            "track": {
                "mode": 0,
                "delay": 300,
                "paths": []
            },
            "dump": True,
            "rpU": ""
        }
        b[16] = start_time
        b[18] = 44

        b[20] = (b[16] >> 24) & 255
        b[21] = (b[16] >> 16) & 255
        b[22] = (b[16] >> 8) & 255
        b[23] = b[16] & 255
        b[24] = (b[16] // (256**4)) & 255
        b[25] = (b[16] // (256**5)) & 255

        b[26] = (Arguments[0] >> 24) & 255
        b[27] = (Arguments[0] >> 16) & 255
        b[28] = (Arguments[0] >> 8) & 255
        b[29] = Arguments[0] & 255

        b[30] = (Arguments[1] // 256) & 255
        b[31] = (Arguments[1] % 256) & 255
        b[32] = (Arguments[1] >> 24) & 255
        b[33] = (Arguments[1] >> 16) & 255

        b[34] = (Arguments[2] >> 24) & 255
        b[35] = (Arguments[2] >> 16) & 255
        b[36] = (Arguments[2] >> 8) & 255
        b[37] = Arguments[2] & 255

        b[38] = url_search_params_list[21]
        b[39] = url_search_params_list[22]

        b[40] = cus[21]
        b[41] = cus[22]

        b[42] = ua[23]
        b[43] = ua[24]

        b[44] = (b[10] >> 24) & 255
        b[45] = (b[10] >> 16) & 255
        b[46] = (b[10] >> 8) & 255
        b[47] = b[10] & 255
        b[48] = b[8]
        b[49] = (b[10] // (256**4)) & 255
        b[50] = (b[10] // (256**5)) & 255

        b[52] = (b[15]['pageId'] >> 24) & 255
        b[53] = (b[15]['pageId'] >> 16) & 255
        b[54] = (b[15]['pageId'] >> 8) & 255
        b[55] = b[15]['pageId'] & 255

        b[57] = b[15]['aid'] & 255
        b[58] = (b[15]['aid'] >> 8) & 255
        b[59] = (b[15]['aid'] >> 16) & 255
        b[60] = (b[15]['aid'] >> 24) & 255

        window_env_list = list(window_env_str.encode('utf-8'))
        b[64] = len(window_env_list)
        b[65] = b[64] & 255
        b[66] = (b[64] >> 8) & 255

        b[69] = 0
        b[70] = b[69] & 255
        b[71] = (b[69] >> 8) & 255

        b[72] = (b[18] ^ b[20] ^ b[26] ^ b[30] ^ b[38] ^ b[40] ^ b[42] ^ b[21] ^ b[27] ^ b[31] ^ b[35] ^ b[39] ^ b[41] ^ b[43] ^ b[22] ^
                 b[28] ^ b[32] ^ b[36] ^ b[23] ^ b[29] ^ b[33] ^ b[37] ^ b[44] ^ b[45] ^ b[46] ^ b[47] ^ b[48] ^ b[49] ^ b[50] ^ b[24] ^
                 b[25] ^ b[52] ^ b[53] ^ b[54] ^ b[55] ^ b[57] ^ b[58] ^ b[59] ^ b[60] ^ b[65] ^ b[66] ^ b[70] ^ b[71])

        bb = [
            b[18], b[20], b[52], b[26], b[30], b[34], b[58], b[38], b[40], b[53], b[42], b[21], b[27], b[54], b[55], b[31],
            b[35], b[57], b[39], b[41], b[43], b[22], b[28], b[32], b[60], b[36], b[23], b[29], b[33], b[37], b[44], b[45],
            b[59], b[46], b[47], b[48], b[49], b[50], b[24], b[25], b[65], b[66], b[70], b[71]
        ]
        bb.extend(window_env_list)
        bb.append(b[72])
        
        bb_str = "".join(map(chr, bb))
        return self.rc4_encrypt(bb_str, "\x79")

    def sign(self, url_search_params, user_agent, arguments):
        result_str = self.generate_random_str() + self.generate_rc4_bb_str(
            url_search_params,
            user_agent,
            "1536|747|1536|834|0|30|0|0|1536|834|1536|864|1525|747|24|24|Win32",
            "cus",
            arguments
        )
        return self.result_encrypt(result_str, "s4") + "="

    def sign_datail(self, params, userAgent):
        return self.sign(params, userAgent, [0, 1, 14])

    def sign_reply(self, params, userAgent):
        return self.sign(params, userAgent, [0, 1, 8])
